arraycontext/impl/pytato/utils.py

```diff
@@ -0,0 +1,97 @@
+__copyright__ = """
+Copyright (C) 2021 University of Illinois Board of Trustees
+"""
+
+__license__ = """
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+
+import numpy as np
+from typing import Any, Dict, Set, Tuple, Mapping
+from pytato.array import SizeParam, Placeholder as BasePlaceholder
+from pytato.array import ShapeType, TagsType, Array, DataWrapper
+from pytato.transform import CopyMapper
+from pytools import UniqueNameGenerator
+
+
+class Placeholder(BasePlaceholder):
+    """
+    An equality-comparable :class:`pytato.Placeholder`.
+    """
+    def __init__(self, name: str,
+                 shape: ShapeType,
+                 dtype: np.dtype,
+                 tags: TagsType = frozenset()
+                 ) -> None:
+        super().__init__(name, shape, dtype, tags)
+
+    def __eq__(self, other) -> bool:
+        return Array.__eq__(self, other)
+
+    def __hash__(self) -> int:
+        return Array.__hash__(self)
```

Comment:
Could you explain this? Isn't the base `PlaceHolder` already equality-comparable?

Reply:
But not in a persistent manner. The equality checks are essentially identity checks https://github.com/inducer/pytato/blob/8f6a07fe041bc5a4bf35e770005f1b25f9b2efb0/pytato/array.py#L1436-L1440.

Reply:
OIC. Question back at you then: Why did we make that choice in `InputArgumentBase`? It seems counterproductive. For a `DataWrapper`, maybe I could see that... but `Placeholder` and such shouldn't really do that.

Reply:
Although I agree that making Placeholders equality comparable seems to be the right thing to do, there are some practical limitations to it. As, all our mappers are "cached" and there would be no way practically feasible way to determine illegal DAGs like
```python
a = pt,make_placeholder("foo", (10, 4))
b = pt,make_placeholder("foo", (10, 4)) 
my_illegal_dag = a + b # unidentical placeholders with same name is most-likely a user-error that should be caught
```

Reply:
> practically feasible way

I think caching the results with `id` for a mapper isn't too bad, we can do this. I don't see any big problems with this.

Reply:
Since, this is more pytato-specific, I vote to move this discussion there. See https://github.com/inducer/pytato/pull/123. Please un-resolved if anyone disagrees.

Reply:
Given that https://github.com/inducer/pytato/pull/123 is now (as good as) in, could this be deleted?

Reply:
Yep, removed.